src/app/shared/interceptors/request-id.interceptor.ts

```diff
@@ -0,0 +1,41 @@
+import { Inject, Injectable } from '@angular/core';
+
+import {
+  HttpEvent,
+  HttpHandler,
+  HttpInterceptor,
+  HttpRequest
+} from '@angular/common/http';
+
+import { Observable } from 'rxjs';
+
+import { WIT_API_URL } from 'ngx-fabric8-wit';
+import { AUTH_API_URL } from 'ngx-login-client';
+import uuid from 'uuid';
+
+
+@Injectable()
+export class RequestIdInterceptor implements HttpInterceptor {
+
+  constructor(
+    @Inject(WIT_API_URL) private witApiUrl: string,
+    @Inject(AUTH_API_URL) private authApiUrl: string
+  ) {}
+
+  intercept(request: HttpRequest<any>, next: HttpHandler): Observable<HttpEvent<any>> {
+    const url = request.url;
+
+    // Attach a X-Request-Id to all requests that don't have one
+    if (url.startsWith(this.witApiUrl) || url.startsWith(this.authApiUrl)) {
+      if (!request.headers.has('X-Request-Id')) {
+        request = request.clone({
```

Comment:
I'm not sure you can do this. I think the `request` is immutable and you are writing back to it. 

When I run the application it has some 401 errors with malformed headers. I'm not sure if this is caused by this change or if it is old.

Reply:
@joshuawilson Yes, `request` is immutable and that is wy we clone it and mutate the cloned object rather than the original. This is how the example on angular.io adds header in interceptor.

Reply:
@joshuawilson The 401s might not be related to this.

Reply:
You close the `request` and then write it back to the same `request`. The examples create a new `const`, they don't use the `request`.

Reply:
Updated